### PR TITLE
Transaction Message Version Bug

### PR DESCRIPTION
**Pull request by braydonf**

### CODE REVIEW

**Comment by braydonf** [comment]
Ready for review!

File: lib/messages/commands/tx.js [hunk 1]
```diff
@@ -2,28 +2,30 @@
...
   this.Transaction = options.Transaction;
-
-  var transaction;
-  if(options instanceof this.Transaction) {
-    transaction = options;
-  } else {
-    transaction = options.transaction;
+  $.checkArgument(
+    _.isUndefined(arg) || arg instanceof this.Transaction,
+    'An instance of Transaction or undefined is expected'
```
Comment (maraoz) [inline lib/messages/commands/tx.js]: why allow `undefined` for all `arg` parameters?

File: lib/messages/message.js [hunk 2]
```diff
@@ -12,7 +12,7 @@ var Hash = bitcore.crypto.Hash;
  * @param {Number=} options.magicNumber
  * @constructor
  */
-function Message(options) {
+function Message(arg, options) {
   this.command = options.command;
   this.magicNumber = options.magicNumber;
 }
```
Comment (maraoz) [inline lib/messages/message.js]: why did you add the `arg` here if it's never used?

File: test/messages/message.js [hunk 3]
```diff
@@ -16,7 +16,7 @@ describe('Message', function() {
 
   describe('#toBuffer', function() {
     it('serialize to a buffer', function() {
-      var message = new Message({magicNumber: 0xd9b4bef9, command: 'command'});
+      var message = new Message(null, {magicNumber: 0xd9b4bef9, command: 'command'});
```
Comment (maraoz) [inline test/messages/message.js]: I don't see the need for this extra `null`, maybe remove the first parameter? (see previous comment)

File: lib/messages/message.js (hunk 2 above)
Comment (braydonf) [inline lib/messages/message.js]: So that it matches the same order of arguments of the inherited messages, but it may not be necessary.

File: lib/messages/commands/tx.js (hunk 1 above)
Comment (braydonf) [inline lib/messages/commands/tx.js]: Mostly because of this test: https://github.com/braydonf/bitcore-p2p/blob/bug/transactions-message/test/messages/index.js#L58 I'll take a look and see if it's necessary.

Comment (maraoz) [inline lib/messages/commands/tx.js]: awesome, thanks!

Comment (braydonf) [inline lib/messages/commands/tx.js]: It's necessary to be able to construct a message without the `arg` set.

There is also another test here that constructs messages without the arg: https://github.com/bitpay/bitcore-p2p/blob/master/test/messages/index.js#L89

More importantly though a message needs to be able to be constructed partially the options only, and then modified with a payload using those options: https://github.com/braydonf/bitcore-p2p/blob/bug/transactions-message/lib/messages/builder.js#L77

File: lib/messages/commands/addr.js [hunk 4]
```diff
@@ -4,21 +4,31 @@ var Message = require('../message');
...
   this.command = 'addr';
   this.magicNumber = options.magicNumber;
-  this.addresses = options.addresses;
+  $.checkArgument(
+    _.isUndefined(arg) ||
+      (Array.isArray(arg) &&
+       !_.isUndefined(arg[0].services) &&
+       !_.isUndefined(arg[0].ip) &&
+       !_.isUndefined(arg[0].port)),
+    'First argument is expected to be an array of addrs'
```
Comment (eordano) [inline lib/messages/commands/addr.js]: Or `undefined`?

File: lib/messages/commands/headers.js [hunk 5]
```diff
@@ -18,12 +20,16 @@ var $ = bitcore.util.preconditions;
...
-function HeadersMessage(options) {
+function HeadersMessage(arg, options) {
   Message.call(this, options);
   this.BlockHeader = options.BlockHeader;
   this.magicNumber = options.magicNumber;
   this.command = 'headers';
-  this.headers = options.headers;
+  $.checkArgument(
+    _.isUndefined(arg) || (Array.isArray(arg) && arg[0] instanceof this.BlockHeader),
+    'First argument is expected to be an array of BlockHeader instances'
```
Comment (eordano) [inline lib/messages/commands/headers.js]: ditto, or `undefined`?

**Comment by eordano** [comment]
LGTM. I think that either take out "undefined" from all `$.checkArgument`'s messages or add it to all of them (just a nit)
